### DISCUSSION TITLE
Various 386 models

### DISCUSSION THREAD

**Discussion by Torinde** [bochs-emu/Bochs#303] [Q&A] [answered]
Which of the [discontinued 386 and 387](https://en.wikipedia.org/wiki/List_of_discontinued_x86_instructions) instructions are emulated by Bochs? (mostly System Management Mode related)
  - [ ] i386
  - [ ] Am386
  - [ ] C&T 386
  - [ ] Cyrix 386
  - [ ] ALi 386
  - [ ] Intel 387/287XL
  - [ ] IIT 387
  - [ ] Cyrix 387

**Comment by stlintel** [comment, marked as answer]
None of them

**Reply by Torinde** [reply]
FWIW, two of those are present in Intel 386/486/Quark and Cyrix CPUs. Intel Quark was discontinued only recently.
- MOV r32,TRx	0F 24 /r	Move from test register	
- MOV TRx,r32	0F 26 /r	Move to test register

**Comment by Torinde** [comment]
Thanks for the quick response!

Separately, looking at [cpuid.cc](https://github.com/bochs-emu/Bochs/blob/master/bochs/cpu/cpuid.cc) and [CHANGES](https://github.com/bochs-emu/Bochs/blob/REL_2_8_FINAL/bochs/CHANGES)
1. MONITORX is mentioned as implemented in CHANGES, but I don't see it in `cpuid.cc`?
2. MCOMMIT, INVLPGB, PTWRITE, PREFETCHWT1, TDX/SEAM, SEV-ES/VMGEXIT, SEV-SNP/RMP - I didn't find them - are some of those implemented?
3. VIA PadLock (multiple extensions) - I think it'll be good to mention those as "not implemented" in cpuid.cc, on the correct place for #304
4. [CL1INVMB](https://en.wikipedia.org/wiki/X86_instruction_listings#Undocumented_instructions_that_appear_only_in_a_limited_subset_of_x86_CPUs_include) (SCC Single-chip Cloud Computer), [LRBni](http://tomforsyth1000.github.io/papers/LRBNI%20origins%20v4%20full%20fat.pdf) (Larrabee/Knights Ferry KNF), LRBni2 (Knights Corner KNC) - I assume you don't plan on adding those

**Reply by stlintel** [reply]
MONITORX is AMD feature and AMD's CPUID are quite behind. 
It should be enabled in some corresponding CPUID model - the newest we have for AMD is Ryzen.
As I understood Ryzen yet has no MONITORX so for MONITORX a new model has to be introduced.

MCOMMIT, INVLPGB, PTWRITE, PREFETCHWT1, TDX/SEAM, SEV-ES/VMGEXIT, SEV-SNP/RMP - None of these AMD extensions are implemented or planned to be implemented. Mainly because of lack of testing. I have no AMD machine or reference simulator which implements them correctly already.
Also security/encryption stuff like TDX, SGX and similar won't be implemented because I am not dealing with secure firmware these things come with.

VIA PadLock - relevant only for VIA CPU. If we want to add some VIA CPU model to Bochs (which one?) you may try to list all the gaps and we might build a plan how to close them. Not all the gaps have to be closed immediately, even for most recent Intel models there are blindspots which are not urgent or necessary.

Yes, I am not planning to do LRB and related stuff yet.

**Reply by Torinde** [reply]
I understand that most of those are not a priority.
My point was just that the bits for all of those (and other known) instruction (sets) **to be mentioned/commented in cpuid.cc**, either as:
- implemented
- not implemented

[`MONITORX`](https://en.wikipedia.org/wiki/CPUID#EAX=80000001h:_Extended_Processor_Info_and_Feature_Bits): According what's in Wikipedia those are supported from Excavator onwards. I think it's present for example on [Ryzen 5 1500X](http://users.atw.hu/instlatx64/AuthenticAMD/AuthenticAMD0800F11_K17_Zen3_CPUID.txt), which has CPUID 80000001: 00800F11-20000000-35C233FF-2FD3FBFF (third group, bit29 in reverse order?). It's already implemented in [Bochs 2.6.9](https://github.com/bochs-emu/Bochs/blob/9177c93e4e69a5c9408a1393766f882ed1313ab3/bochs/CHANGES#L500)

`MCOMMIT` (Zen2), `INVLPGB` (Zen3) - which is your Ryzen model?
`PTWRITE` - it's an Intel one - and I think you have a [note about it](https://github.com/bochs-emu/Bochs/blob/master/bochs/cpu/todo): "No plans for ..., Intel Processor Trace"
`PREFETCHWT1` (Knights Landing, YongFeng), `TDX/SEAM` (Sapphire Rapids), probably you don't have any of those yet.

VIA `PadLock` - I try to list the gaps on its Issue: #312
